lib/cloud_controller/upload_buildpack.rb

```diff
@@ -47,8 +50,31 @@ def upload_buildpack(buildpack, bits_file_path, new_filename)
       true
     end
 
+    def extract_stack_from_buildpack(bits_file_path)
+      bits_file_path = bits_file_path.path if bits_file_path.respond_to?(:path)
+      output, _, status = Open3.capture3('unzip', '-p', bits_file_path, 'manifest.yml')
+      YAML.safe_load(output).dig('stack') if status.success?
+    end
+
     private
 
+    def determine_new_stack(buildpack, bits_file_path)
+      extracted_stack = extract_stack_from_buildpack(bits_file_path)
+      if extracted_stack.to_s != '' && !Stack.where(name: extracted_stack).first
+        raise CloudController::Errors::ApiError.new_from_details('BuildpackStackDoesNotExist', extracted_stack)
+      end
+      new_stack = [extracted_stack, buildpack.stack, Stack.default.name].find { |s| s.to_s != '' && s.to_s != 'unknown' }
+      if buildpack.stack != 'unknown' && buildpack.stack != new_stack
+        raise CloudController::Errors::ApiError.new_from_details('BuildpackStacksDontMatch', new_stack, buildpack.stack)
+      end
+
+      if buildpack.stack != new_stack && Buildpack.find(name: buildpack.name, stack: new_stack)
+        raise CloudController::Errors::ApiError.new_from_details('BuildpackNameStackTaken', buildpack.name, new_stack)
```

Comment:
This feels like duplicate logic for the model validation that checks for unique (buildpack, stack) pairs.

I've also noticed that we currently don't raise anywhere in this file currently (and explicitly catch Sequel errors). I suspect that we don't have a good reason for doing this, but it might make sense to think about how we handle exceptions in this file.